core/lib/zksync_core/src/api_server/tx_sender/proxy.rs

```diff
@@ -25,20 +113,41 @@ impl TxProxy {
         let client = HttpClientBuilder::default().build(main_node_url).unwrap();
         Self {
             client,
-            tx_cache: RwLock::new(HashMap::new()),
+            tx_cache: TxCache::default(),
         }
     }
 
     pub async fn find_tx(&self, tx_hash: H256) -> Option<L2Tx> {
-        self.tx_cache.read().await.get(&tx_hash).cloned()
+        self.tx_cache.get_tx(tx_hash).await
     }
 
     pub async fn forget_tx(&self, tx_hash: H256) {
-        self.tx_cache.write().await.remove(&tx_hash);
+        self.tx_cache.remove_tx(tx_hash).await;
+    }
+
+    pub async fn save_tx(&self, tx: L2Tx) {
+        self.tx_cache.push(tx).await;
+    }
+
+    pub async fn get_nonces_by_account(&self, account_address: Address) -> Vec<Nonce> {
+        self.tx_cache.get_nonces_for_account(account_address).await
     }
 
-    pub async fn save_tx(&self, tx_hash: H256, tx: L2Tx) {
-        self.tx_cache.write().await.insert(tx_hash, tx);
+    pub async fn next_nonce_by_initiator_account(
+        &self,
+        account_address: Address,
+        current_nonce: u32,
+    ) -> Nonce {
+        let mut expected_nonce = current_nonce.saturating_sub(1);
```

Comment:
Dumb question: Shouldn't this use the same logic as in `next_nonce_by_initiator_account()`, i.e.:

```rust
let mut expected_nonce = current_nonce;
// ...
if pending_nonce.0 == expected_nonce {
    expected_nonce += 1;
}
Nonce(expected_nonce)
```

I *think* these two approaches are always equivalent (including edge cases when `current_nonce == 0` and/or `nonces` start with 0), so I'd prefer using the same approach as in another place (which is tested / has been used in production).

Reply:
@Deniallugo @slowli is this resolved?

Reply:
As far as I see it's